internal/pipe/krew/krew.go

```diff
@@ -0,0 +1,325 @@
+	}
+
+	filters := []artifact.Filter{
+		artifact.Or(
+			artifact.ByGoos("darwin"),
+			artifact.ByGoos("linux"),
+			artifact.ByGoos("windows"),
+		),
+		artifact.Or(
+			artifact.ByGoarch("amd64"),
+			artifact.ByGoarch("arm64"),
+			artifact.ByGoarch("all"),
+			artifact.And(
+				artifact.ByGoarch("arm"),
+				artifact.ByGoarm(krew.Goarm),
+			),
+		),
+		artifact.ByType(artifact.UploadableArchive),
+	}
+	if len(krew.IDs) > 0 {
+		filters = append(filters, artifact.ByIDs(krew.IDs...))
+	}
+
+	archives := ctx.Artifacts.Filter(artifact.And(filters...)).List()
+	if len(archives) == 0 {
+		return ErrNoArchivesFound
+	}
+
+	name, err := tmpl.New(ctx).Apply(krew.Name)
+	if err != nil {
+		return err
+	}
+	krew.Name = name
+
+	content, err := buildPlugin(ctx, krew, cl, archives)
+	if err != nil {
+		return err
+	}
+
+	filename := krew.Name + ".yml"
+	yamlPath := filepath.Join(ctx.Config.Dist, filename)
+	log.WithField("plugin", yamlPath).Info("writing")
+	if err := os.WriteFile(yamlPath, []byte(content), 0o644); err != nil { //nolint: gosec
+		return fmt.Errorf("failed to write krew plugin: %w", err)
+	}
+
+	ctx.Artifacts.Add(&artifact.Artifact{
+		Name: filename,
+		Path: yamlPath,
+		Type: artifact.KrewPlugin,
+		Extra: map[string]interface{}{
+			krewConfigExtra: krew,
+		},
+	})
+
+	return nil
+}
+
+func buildPlugin(ctx *context.Context, krew config.Krew, client client.Client, artifacts []*artifact.Artifact) (string, error) {
+	data, err := dataFor(ctx, krew, client, artifacts)
+	if err != nil {
+		return "", err
+	}
+	return doBuildPlugin(data)
+}
+
+func doBuildPlugin(data Plugin) (string, error) {
+	out, err := yaml.Marshal(data)
+	if err != nil {
+		return "", fmt.Errorf("krew: failed to marshal yaml: %w", err)
+	}
+	return string(out), nil
+}
+
+func dataFor(ctx *context.Context, cfg config.Krew, cl client.Client, artifacts []*artifact.Artifact) (Plugin, error) {
+	desc, err := tmpl.New(ctx).Apply(cfg.Description)
+	if err != nil {
+		return Plugin{}, err
+	}
+
+	shortDesc, err := tmpl.New(ctx).Apply(cfg.ShortDescription)
+	if err != nil {
+		return Plugin{}, err
+	}
+
+	result := Plugin{
+		APIVersion: apiVersion,
+		Kind:       kind,
+		Metadata: Metadata{
+			Name: cfg.Name,
+		},
+		Spec: Spec{
+			Homepage:         cfg.Homepage,
+			Version:          ctx.Version,
+			ShortDescription: shortDesc,
+			Description:      desc,
+			Caveats:          cfg.Caveats,
+		},
+	}
+
+	for _, art := range artifacts {
+		sum, err := art.Checksum("sha256")
+		if err != nil {
+			return result, err
+		}
+
+		if cfg.URLTemplate == "" {
+			url, err := cl.ReleaseURLTemplate(ctx)
+			if err != nil {
+				return result, err
+			}
+			cfg.URLTemplate = url
+		}
+		url, err := tmpl.New(ctx).WithArtifact(art, map[string]string{}).Apply(cfg.URLTemplate)
+		if err != nil {
+			return result, err
+		}
+
+		goarch := []string{art.Goarch}
+		if art.Goarch == "all" {
+			goarch = []string{"amd64", "arm64"}
+		}
+
+		for _, arch := range goarch {
```

Comment:
Would this work also with eg. cosign binary? As you now they have this multibinaries with same Arch/OS combination

Reply:
from what I looked into their code/index, seems like only one binary is allowed... maybe we should even hardfail on those cases